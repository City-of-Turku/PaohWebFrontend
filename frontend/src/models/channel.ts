import { Language } from '../types';

export const channelTypes = [
  'EChannel',
  'Phone',
  'PrintableForm',
  'ServiceLocation',
  'WebPage',
] as const;

export type ChannelType = typeof channelTypes[number];

export const isChannelType = (val: string): val is ChannelType => {
  // widen channelTypes to string[] so indexOf(x) works
  return (channelTypes as readonly string[]).indexOf(val) !== -1;
};

export type IChannelDescType = 'Description' | 'Summary';

export interface IChannelDescription {
  value: string;
  type: IChannelDescType;
  language: Language;
}

export interface IPhoneNumber {
  number: string;
  prefixNumber?: string;
  chargeDescription?: string;
  serviceChargeType?: string;
}

export interface IAddress {
  type: string;
  subtype: string;
  streetNumber?: string;
  postalCode?: string;
  latitude?: string;
  longitude?: string;
  streetName?: string;
  postOffice?: string;
  municipalityCode?: string;
  municipalityName?: string;
}

export interface IChannelUrl {
  url: string;
  type: string;
}

export interface IChannel {
  id: string;
  type: ChannelType;
  name: string;
  descriptions: IChannelDescription[];
  webPages: string[];
  emails: string[];
  phoneNumbers: IPhoneNumber[];
  addresses: IAddress[];
  channelUrls: IChannelUrl[];
}
