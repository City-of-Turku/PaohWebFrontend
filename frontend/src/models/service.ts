import { Language } from '../types';

export type ServiceDescriptionType =
  | 'ChargeTypeAdditionalInfo'
  | 'DeadLine'
  | 'Description'
  | 'GD_Description'
  | 'GD_GeneralDescriptionTypeAdditionalInformation'
  | 'GD_Summary'
  | 'ProcessingTime'
  | 'ServiceType'
  | 'Summary'
  | 'UserInstruction'
  | 'ValidityTime';

export interface IDescription {
  value: string;
  type: ServiceDescriptionType;
  language: Language;
  autoTranslated: boolean;
  original?: string; // If name was automatically translated, this contains the original (Finnish)
  originalLanguage?: string;
}

export interface IOrganization {
  id: string | null;
  name: string;
  roleType: OrgRole;
}

export enum OrgRole {
  Responsible = 'Responsible',
  Producer = 'Producer',
  OtherResponsible = 'OtherResponsible',
}

export interface IServiceArea {
  name: string;
  type: string;
  code: string;
}

export interface IServiceName {
  value: string;
  language: Language;
  autoTranslated: boolean;
  original?: string; // If name was automatically translated, this contains the original (Finnish)
  originalLanguage?: Language;
}

export interface IService {
  id: string;
  name: IServiceName;
  descriptions: IDescription[];
  organizations: IOrganization[];
  areas: IServiceArea[];
}
