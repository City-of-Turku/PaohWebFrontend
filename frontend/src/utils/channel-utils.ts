import {
  IChannel,
  IChannelDescription,
  IPhoneNumber,
  IAddress,
} from 'models/channel';

function formatAddressParts(address: IAddress): string[] {
  const streetAddress = address.streetName
    ? `${address.streetName} ${address.streetNumber}`
    : '';
  const municipality = address.municipalityName
    ? `${address.postalCode} ${address.municipalityName}`
    : '';

  return [streetAddress, municipality];
}

function formatAddressOneliner(address: IAddress): string {
  return formatAddressParts(address).join(', ');
}

function formatAddressMultiline(address: IAddress): string {
  return formatAddressParts(address).join('\n');
}

function formatPhoneNumber(number: IPhoneNumber): string {
  return `${number.prefixNumber || ''}${number.number}`;
}

function generateMapsDirectionLink(address: string): string {
  const mapsBaseUrl = 'https://www.google.com/maps/dir/?api=1';
  const destination = '' + encodeURIComponent(address);
  return `${mapsBaseUrl}&destination=${destination}`;
}

function generateMapsLocationLink(longitude: string, latitude: string): string {
  return `https://www.openstreetmap.org/?mlat=${latitude}&mlon=${longitude}&zoom=16`;
}

function getChannelDescription(channel: IChannel): {
  description: string;
  descLang: string | undefined;
} {
  const summary = channel.descriptions.find((d) => d.type == 'Summary');
  const description = channel.descriptions.find(
    (d: IChannelDescription) => d.type == 'Description',
  );

  const chosen = summary || description;

  return {
    description: chosen ? chosen.value : '',
    descLang: chosen ? chosen.language : undefined,
  };
}

function uniqueFormattedAddresses(
  addresses: IAddress[],
  formattingFunc: (a: IAddress) => string,
): string[] {
  const uniqueAddresses: string[] = addresses
    .filter((address: IAddress) => address.type === 'Location')
    .map((a) => formattingFunc(a))
    .filter((formattedAdd) => formattedAdd !== '')
    .filter((addr, index, self) => self.indexOf(addr) === index);

  return uniqueAddresses;
}

export interface IFormmattedLocation {
  address: string;
  latitude: string | undefined;
  longitude: string | undefined;
}

function uniqueFormattedLocationAddresses(
  addresses: IAddress[],
  formattingFunc: (a: IAddress) => string,
): IFormmattedLocation[] {
  const uniqueAddresses: IFormmattedLocation[] = addresses
    .filter((address: IAddress) => address.type === 'Location')
    .map((a) => {
      return {
        address: formattingFunc(a),
        latitude: a.latitude,
        longitude: a.longitude,
      };
    })
    .filter((formattedAdd) => formattedAdd.address !== '')
    .filter(
      (addr, index, self) =>
        self.findIndex((a) => a.address === addr.address) === index,
    );

  return uniqueAddresses;
}

function uniqueFormattedPhoneNumbers(numbers: IPhoneNumber[]): string[] {
  const uniqueNumbers: string[] = numbers
    .map((n) => formatPhoneNumber(n))
    .filter((num, index, self) => self.indexOf(num) === index);
  return uniqueNumbers;
}

export {
  formatAddressMultiline,
  formatAddressOneliner,
  formatPhoneNumber,
  generateMapsDirectionLink,
  generateMapsLocationLink,
  getChannelDescription,
  uniqueFormattedAddresses,
  uniqueFormattedLocationAddresses,
  uniqueFormattedPhoneNumbers,
};
