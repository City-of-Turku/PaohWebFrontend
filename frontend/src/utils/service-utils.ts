import {
  IDescription,
  IOrganization,
  IServiceArea,
  OrgRole,
} from 'models/service';

/* Check if string is valid UUID */
function checkIfValidUUID(str: string): boolean {
  const regexExp =
    /^[0-9a-f]{8}\b-[0-9a-f]{4}\b-[0-9a-f]{4}\b-[0-9a-f]{4}\b-[0-9a-f]{12}$/gi;
  return regexExp.test(str.toLowerCase());
}

/**
 * Order of descriptions to use: Description > GD_Description > Summary > GD_Summary
 * If none of them exist, return an empty description
 */
function getDescription(
  descriptions: IDescription[],
): IDescription | undefined {
  // Convert type: x, value: y structure into {x: y} meaning description types serve as keys for values (descriptions)
  const mappedDescriptions = descriptions.map((desc: IDescription) => ({
    [desc.type]: {
      value: desc.value,
      language: desc.language,
      autoTranslated: desc.autoTranslated,
      ...(desc.original && {
        original: desc.original,
        originalLanguage: desc.originalLanguage,
      }),
    },
  }));
  const descData = Object.assign({}, ...mappedDescriptions);
  const descriptionOrder = [
    'Description',
    'Summary',
    'GD_Description',
    'GD_Summary',
  ];

  const selectedType = descriptionOrder.find(
    (descName: string) =>
      descData[descName] &&
      descData[descName].value !== null &&
      descData[descName].value !== undefined &&
      descData[descName].value !== '',
  );

  if (selectedType) {
    return descData[selectedType];
  }
  return undefined;
}

/**
 *
 * @param areas
 * @param areasText
 * @returns
 */
function getLocationSubheader(
  areas: IServiceArea[],
  areasText: string,
): string | null {
  if (areas && areas.length > 0) {
    return areas.length === 1 ? areas[0].name : `${areas.length} ${areasText}`;
  }
  return null;
}

/**
 * Get name of (one) organization responsible for a service
 * @param organizations Array of organizations
 * @returns name of (first) organization with role Responsible - or if such is not found, returns null
 */
function getOrganizationName(organizations: IOrganization[]): string | null {
  const responsible = organizations.filter(
    (o) => o.roleType === OrgRole.Responsible,
  );
  if (responsible.length > 0) {
    return responsible[0].name;
  }
  return null;
}

export {
  checkIfValidUUID,
  getDescription,
  getLocationSubheader,
  getOrganizationName,
};
