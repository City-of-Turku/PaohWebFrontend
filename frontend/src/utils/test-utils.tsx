import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import { ThemeProvider as EmotionThemeProvider } from '@emotion/react';
import { ThemeProvider as MuiThemeProvider } from '@material-ui/core';
import { theme } from '../theme/theme';
import { getTranslations } from 'translations/locales';

interface ProviderProps {
  children?: React.ReactNode;
}

function AllTheProviders({ children }: ProviderProps) {
  return (
    <IntlProvider locale="fi" messages={getTranslations('fi')}>
      <MuiThemeProvider theme={theme}>
        <EmotionThemeProvider theme={theme}>{children}</EmotionThemeProvider>
      </MuiThemeProvider>
    </IntlProvider>
  );
}

// eslint-disable-next-line
const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';

export { customRender as render };

/**
 * According to the React docs, Hooks cannot be called from anywhere like regular javascript functions.
 * They can only be called from a functional component or from other hooks.
 *
 * The component below can be used to call hooks in tests. Using a callback, we can pass
 * the result of our custom hook in the outside scope where we want to test it.
 */

/* eslint-disable  @typescript-eslint/no-explicit-any */
const TestHook = ({ callback }: any): null => {
  callback();
  return null;
};

/* eslint-disable  @typescript-eslint/no-explicit-any */
export const testHook = (callback: any): void => {
  render(<TestHook callback={callback} />);
};
