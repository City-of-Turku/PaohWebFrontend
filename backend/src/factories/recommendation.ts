import { faker } from '@faker-js/faker';
import { makeFakeService } from './service';
import { makeFakeChannel } from './channel';
import { RecommendationDTO } from '../dtos/recommendation.dto';
import { generateList } from './factory-utils';

export const makeFakeRecommendationList = (
  min = 1,
  max = 3,
): RecommendationDTO[] => {
  return generateList(makeFakeRecommendation(), min, max);
};

export const makeFakeRecommendation = (
  channelCount?: number,
): RecommendationDTO => {
  const minChannels = channelCount || 1;
  const maxChannels = channelCount || 4;
  return {
    service: makeFakeService(),
    channels: generateList(makeFakeChannel(), minChannels, maxChannels),
    score: faker.datatype.boolean() ? faker.datatype.number(1) : undefined,
  };
};
