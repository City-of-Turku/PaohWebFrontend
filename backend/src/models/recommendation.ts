import { IService } from './service';
import { IChannel } from './channel';

export interface IRecommendation {
  service: IService;
  channels: IChannel[];
  score?: number;
}
