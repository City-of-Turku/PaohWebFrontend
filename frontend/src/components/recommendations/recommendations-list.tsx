/** @jsxRuntime classic */
/** @jsx jsx */
import { css, jsx } from '@emotion/react';
import { FC, Fragment, useRef } from 'react';
import { CircularProgress, Typography, useTheme } from '@material-ui/core';
import { observer } from 'mobx-react-lite';
import { IRecommendation } from 'models/recommendation';
import { ServiceCard } from '../service-card/service-card';
import { useStore } from 'stores/store-hooks';
import styled from '@emotion/styled';
import { useIntl } from 'react-intl';
import { StoreState } from 'stores/store-state';
import { ScrollToButton } from '../buttons/scroll-to-button';

interface IRecommendationsList {
  recommendations: IRecommendation[];
}

export const RecommendationsList: FC<IRecommendationsList> = observer(
  ({ recommendations }) => {
    const intl = useIntl();
    const theme = useTheme();
    const recommendationsStore = useStore('recommendationsStore');

    const scrollRef = useRef(null);
    const bottomRef = useRef(null);

    if (recommendationsStore.state == StoreState.pending) {
      return (
        <LoadingRoot>
          <CircularProgress size={50} />
        </LoadingRoot>
      );
    }

    return (
      <Fragment>
        {/* Display error message at the top of the recommendations card list  */}
        {recommendationsStore.state == StoreState.error && (
          <Typography
            variant="body1"
            css={css`
              && {
                font-weight: 500;
                color: ${theme.palette.error.dark};
                background-color: ${theme.palette.secondary.main};
                padding-top: 0.25rem;
                padding-bottom: 0.25rem;
                text-align: center;
                margin-top: ${theme.spacing(1)}px;
                margin-bottom: ${theme.spacing(1)}px;
              }
            `}
          >
            {intl.formatMessage({
              id: 'ErrorInRecommendationRetrieval',
              defaultMessage:
                'Valitettavasti suositusten hakeminen ep√§onnistui.',
              description:
                'An explanation to be shown in the recommendations panel if retrieving recommendations from the server failed, for example because connection to the server failed',
            })}
          </Typography>
        )}
        <div
          ref={scrollRef}
          css={css`
            position: relative;
            margin-top: ${theme.spacing(1)}px;
            height: 100%;
            padding-right: ${theme.spacing(1)}px;
            overflow-y: auto;
            > .MuiCard-root {
              margin-bottom: ${theme.spacing(2)}px;
            }

            > div:nth-last-of-type(2) {
              // Last child is the bottom target element
              margin-bottom: ${theme.spacing(0)}px;
            }
          `}
        >
          <div />
          {recommendations.length > 0 ? (
            recommendations.map((rec: IRecommendation) => (
              <ServiceCard
                key={rec.service.id}
                service={rec.service}
                channels={rec.channels}
              />
            ))
          ) : (
            <Typography variant="body1">
              {intl.formatMessage({
                id: 'RecommendationsListEmptyGuidance',
                defaultMessage:
                  'Kerro jotain tilanteestasi saadaksesi suosituksia.',
                description: 'Guidance shown in empty recommendations panel',
              })}
            </Typography>
          )}
          <BottomTarget ref={bottomRef} />
        </div>
        {recommendations.length > 0 && (
          <ScrollToButton
            scrollRef={scrollRef}
            bottomRef={bottomRef}
            autoSwitchDirections={true}
            showText={false}
            bottomMargin={'100px'}
            rightMargin={'3rem'}
          />
        )}
      </Fragment>
    );
  },
);

const BottomTarget = styled.div``;

const LoadingRoot = styled.div`
  display: flex;
  flex-direction: column;
  height: 100%;
  align-items: center;
  justify-content: center;
`;
