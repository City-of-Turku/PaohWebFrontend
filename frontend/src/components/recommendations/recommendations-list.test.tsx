import React from 'react';
import { render, screen } from '../../utils/test-utils';
import { RecommendationsList } from './recommendations-list';
import { IRecommendation } from '../../models/recommendation';
import { MemoryRouter } from 'react-router-dom';

const testRecs: IRecommendation[] = [
  {
    service: {
      id: '1',
      name: {
        value: 'Esimerkkipalvelu',
        language: 'fi',
        autoTranslated: false,
      },
      organizations: [],
      descriptions: [
        {
          value: 'Tämä on palvelun kuvaus.',
          type: 'Description',
          language: 'fi',
          autoTranslated: false,
        },
        {
          value: 'Tämä on tiivistelmä palvelun kuvauksesta.',
          type: 'Summary',
          language: 'fi',
          autoTranslated: false,
        },
      ],
      areas: [],
    },
    channels: [],
  },
];

it('renders empty list correctly', () => {
  const { asFragment } = render(
    <MemoryRouter>
      <RecommendationsList recommendations={[]} />
    </MemoryRouter>,
  );
  expect(asFragment()).toMatchSnapshot();
});

it('renders non-empty list correctly', () => {
  const { asFragment } = render(
    <MemoryRouter>
      <RecommendationsList recommendations={testRecs} />
    </MemoryRouter>,
  );
  expect(asFragment()).toMatchSnapshot();
});

it('shows placeholder text when there are no recommendations', () => {
  render(
    <MemoryRouter>
      <RecommendationsList recommendations={[]} />
    </MemoryRouter>,
  );
  const placeholder = screen.getByText(
    'Kerro jotain tilanteestasi saadaksesi suosituksia.',
  );
  expect(placeholder).toBeInTheDocument();
});

it('does not show placeholder text when recommendations are available', () => {
  render(
    <MemoryRouter>
      <RecommendationsList recommendations={testRecs} />
    </MemoryRouter>,
  );
  const placeholder = screen.queryByText(
    'Kerro jotain tilanteestasi saadaksesi suosituksia.',
  );
  expect(placeholder).toBeNull();
});
