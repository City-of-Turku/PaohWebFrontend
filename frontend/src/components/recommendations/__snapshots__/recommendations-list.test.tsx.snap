// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty list correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1czs6r7"
  >
    <div />
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Kerro jotain tilanteestasi saadaksesi suosituksia.
    </p>
    <div
      class="css-0"
    />
  </div>
</DocumentFragment>
`;

exports[`renders non-empty list correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1czs6r7"
  >
    <div />
    <div
      class="MuiPaper-root MuiCard-root css-1iszrn9 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
          >
            <div
              class="css-oj1ggt"
            >
              <div
                class="css-14yu5aj"
              >
                <h3
                  class="MuiTypography-root organization-name MuiTypography-body1"
                />
                <h2
                  class="MuiTypography-root service-name MuiTypography-h1"
                  lang="fi"
                >
                  <a
                    class="css-shy10a"
                    href="/1"
                  >
                    <span
                      class="link-content"
                    >
                      Esimerkkipalvelu
                    </span>
                  </a>
                </h2>
                <h3
                  class="MuiTypography-root service-area MuiTypography-body1"
                />
              </div>
            </div>
          </span>
        </div>
      </div>
      <div
        class="MuiCardContent-root"
      >
        <div
          class="css-1anmknu"
        >
          <div
            class="description"
            lang="fi"
          >
            Tämä on palvelun kuvaus.
          </div>
        </div>
      </div>
      <div
        class="MuiCardContent-root"
      />
    </div>
    <div
      class="css-0"
    />
  </div>
  <button
    class="MuiButtonBase-root MuiIconButton-root hidden css-lqwbw1-ScrollToButton"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiIconButton-label"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-circle-arrow-up fontawesome-icon"
        color="white"
        data-icon="circle-arrow-up"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 0C114.6 0 0 114.6 0 256c0 141.4 114.6 256 256 256s256-114.6 256-256C512 114.6 397.4 0 256 0zM382.6 254.6c-12.5 12.5-32.75 12.5-45.25 0L288 205.3V384c0 17.69-14.33 32-32 32s-32-14.31-32-32V205.3L174.6 254.6c-12.5 12.5-32.75 12.5-45.25 0s-12.5-32.75 0-45.25l103.1-103.1C241.3 97.4 251.1 96 256 96c4.881 0 14.65 1.391 22.65 9.398l103.1 103.1C395.1 221.9 395.1 242.1 382.6 254.6z"
          fill="currentColor"
        />
      </svg>
    </span>
  </button>
</DocumentFragment>
`;
